libtiledbsoma/src/soma/soma_column.h

```diff
@@ -0,0 +1,343 @@
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ *   This file defines the SOMAColumn class.
+ */
+
+#ifndef SOMA_COLUMN_H
+#define SOMA_COLUMN_H
+
+#include <algorithm>
+#include <any>
+#include <optional>
+#include <string>
+#include <vector>
+
+#include <spdlog/fmt/fmt.h>
+#include "enums.h"
+#include "managed_query.h"
+#include "nanoarrow/nanoarrow.hpp"
+#include "utils/common.h"
+
+namespace tiledbsoma {
+using namespace tiledb;
+
+class SOMAColumn {
+   public:
+    //===================================================================
+    //= public non-static
+    //===================================================================
+
+    /**
+     * Construct a new SOMAColumn object.
+     *
+     * @param ctx TileDB context
+     */
+    SOMAColumn(const Context& ctx)
+        : ctx(ctx){};
+    SOMAColumn(const SOMAColumn&) = default;
+    SOMAColumn(SOMAColumn&&) = default;
+    SOMAColumn& operator=(const SOMAColumn&) = default;
+    SOMAColumn& operator=(SOMAColumn&&) = default;
+
+    /**
+     * Get the SOMAColumn name as defined in schema.
+     */
+    virtual std::string name() const = 0;
+
+    /**
+     * If true, this column is used as index.
+     *
+     * @remark SOMAColumns used as indexes should define at least one TileDB
+     * Dimension
+     */
+    virtual bool isIndexColumn() const = 0;
+
+    /**
+     * Get the TileDB Dimensions defined by the SOMAColumn object, if any.
+     */
+    virtual std::optional<std::vector<Dimension>> tiledb_dimensions() = 0;
+
+    /**
+     * Get the TileDB Attributes defined by the SOMAColumn object, if any.
+     */
+    virtual std::optional<std::vector<Attribute>> tiledb_attributes() = 0;
+
+    /**
+     * Get the TileDB Enumerations used by the SOMAColumn object, if any.
+     */
+    virtual std::optional<std::vector<Enumeration>> tiledb_enumerations() = 0;
+
+    /**
+     * Get the SOMAColumn type. Each subclass should define its own type.
+     */
+    virtual soma_column_datatype_t type() const = 0;
+
+    /**
+     * Get the datatype of the TileDB Dimensions if any. All dimensions must
+     * have the same type.
+     */
+    virtual std::optional<tiledb_datatype_t> domain_type() const = 0;
+
+    /**
+     * Get the datatype of the TileDB Attributes if any. All attributes must
+     * have the same type.
+     */
+    virtual std::optional<tiledb_datatype_t> data_type() const = 0;
+
+    /**
+     * @brief Select columns names to query (dim and attr). If the
+     * `if_not_empty` parameter is `true`, the column will be selected iff the
```

Comment:
"iff" is specialist mathematical jargon that non-mathematicians are unlikely to recognize. Please don't use it.